### DISCUSSION TITLE
Re-rendering and filtering on server components with URLSearchParams

### DISCUSSION THREAD

**Discussion by umutcanuvut** [vercel/next.js#66797] [Help]
### Summary

I am trying to filter and show fetched data on a server component. I am using an external API that doesn't support filtering.

```
// app.page.tsx
const Home = async ({
  searchParams,
}) => {
  const { data, timestamp } = await getData();

  const query = searchParams?.query?.toLowerCase() || "";
  
  const filteredData = data.filter((item) =>
    item.name.toLowerCase().includes(query),
  );
  
  return (
      <div>
        <Search />                        // client component takes the input and update searchParams
        <p>
          Last fetched at: {new Date(timestamp).toLocaleString()}
        </p>
        <Table data={filteredData} />.    // server component rendering filtered data
      </div>
  );
};
```

- With this approach whenever searchParams change all the page is re-rendered and data is re-fetched; which is expected.

- I want to prevent re-fetch but still filter the data and pass to the `<Table>` component.

- Tried using [Native History API](https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#using-the-native-history-api) on `<Search>`, it prevents a new render but it also doesn't filter the data.

- Tried moving getData or filterData or both to `<Table>` still causes fresh fetches on every keystore.

- Also, changing `<Table>` to client component and filtering it there with state might solve the problem. But I'm trying to use it as a server component since it doesn't have any interactivity.

**<ins>What I am trying to achieve is fetch the data once, filter it based on URLParams and keep `<Home>` and `<Table>` as server components.</ins>**

**Comment by sanneh2** [comment]
If you're still doing this. Fetch the initial data once, and then pass it down to a client component. Based on the URL search params you can read the filtering options inside the `Table` you are using and filter the data on the client side and return the new view with `useState`.

What I was thinking about is how to render the initial data on the server and not lose the static benefits of the rendered page.
I think if you have the `Table` a client component but another `TableItem` component inside of it which should make it a server component.  But only if you pass it as props to a client component by following this pattern: https://nextjs.org/docs/app/building-your-application/rendering/composition-patterns#supported-pattern-passing-server-components-to-client-components-as-props
